### ISSUE TITLE
where and how to enable kernel debugging

### ISSUE THREAD

**Issue by bug-developer021** [jupyter/enterprise_gateway#980]
Note: This issue was originally opened with a question about how to configure Lab extensions within a Hub and EG environment.  As that issue was quickly resolved, the discussion migrated to one regarding kernel debugging.  Since this too was an evolving discussion, we've decided to repurpose this issue one regarding kernel-debugger enablement - which starts here: https://github.com/jupyter/enterprise_gateway/issues/980#issuecomment-884922230

Here's the original issue description...

## Description
Hi , i have deployed Jupyter Hub and Enterprise Gateway on Kubernetes.  Jupyter Hub seems like a client, and Enterprise Gateway like a server. Now, i want to install some juputerlab extension for code edit, visualization and machine learning。
Both JHUB user container and JEG kernel container  are running on Kubernetes, so i'm confused about where (JHUB or JEG) and how to install jupyterlab extension。
The following is the information i got when i installed some extension in jupyterlab

```text
This package has indicated that it needs a corresponding server extension. Please contact your Administrator  
```

## Environment

- Enterprise Gateway Version [2.5.0]
- Jupyter Hub Version [1.4.1]
- Platform: [Kubernetes ]

**Comment by bug-developer021** [comment]
it seems like that some extension should deploy on server side and lab side

**Comment by kevin-bates** [comment]
Hi @bug-developer021 - thanks for the question.  I'm not entirely familiar with Jupyter Hub configuration but perhaps if I give a high-level overview of the various component relationships regarding your question, it might be helpful.

Jupyter Hub spawns single-user notebook servers.  Based on the configuration and what appears to be your scenario, these "notebook servers" are actually jupyter lab servers.  Since you're integrating with Enterprise Gateway, each spawned (lab) server is configured to proxy its kernel-based operations to a previously instanced EG server (presumably using the `gateway-url` configuration option) that is exposed as a Kubernetes service.  The EG k8s service is what then launches and manages pod-based kernels on behalf of the lab server (spawned by Jupyter Hub).

Any necessary (lab) server extensions must be configured however Jupyter Hub exposes the configuration of its spawned servers.  As a result, your question should be approached from the standpoint of JHUB and how one would configure extensions (both server and lab extensions) relative to the spawned servers.  The EG integration should remain untouched but does serve as an example of a similar level of configuration so that each spawned (lab) server knows to forward its kernel management to the EG server (i.e., k8s service).

I would imagine you'd need to ensure the necessary extensions have been installed into the single-user target image, then provide a configuration file that indicates the extensions are enabled.  The JHUB server would then spawn _that_ image and you should find your extensions loaded and available.  But, again, the EG side of things would remain untouched.

Since this issue is more about JHUB configuration, I'm going to close it.  However, please feel free to add follow-up questions and any resolutions you find since the latter may be beneficial to others that stumble upon this issue with a similar question.

**Comment by bug-developer021** [comment]
> Hi @bug-developer021 - thanks for the question. I'm not entirely familiar with Jupyter Hub configuration but perhaps if I give a high-level overview of the various component relationships regarding your question, it might be helpful.
> 
> Jupyter Hub spawns single-user notebook servers. Based on the configuration and what appears to be your scenario, these "notebook servers" are actually jupyter lab servers. Since you're integrating with Enterprise Gateway, each spawned (lab) server is configured to proxy its kernel-based operations to a previously instanced EG server (presumably using the `gateway-url` configuration option) that is exposed as a Kubernetes service. The EG k8s service is what then launches and manages pod-based kernels on behalf of the lab server (spawned by Jupyter Hub).
> 
> Any necessary (lab) server extensions must be configured however Jupyter Hub exposes the configuration of its spawned servers. As a result, your question should be approached from the standpoint of JHUB and how one would configure extensions (both server and lab extensions) relative to the spawned servers. The EG integration should remain untouched but does serve as an example of a similar level of configuration so that each spawned (lab) server knows to forward its kernel management to the EG server (i.e., k8s service).
> 
> I would imagine you'd need to ensure the necessary extensions have been installed into the single-user target image, then provide a configuration file that indicates the extensions are enabled. The JHUB server would then spawn _that_ image and you should find your extensions loaded and available. But, again, the EG side of things would remain untouched.
> 
> Since this issue is more about JHUB configuration, I'm going to close it. However, please feel free to add follow-up questions and any resolutions you find since the latter may be beneficial to others that stumble upon this issue with a similar question.

Thanks for your detailed answers! I deploy these extensions into the single-user image, It works!
But  when I have installed the  elyra extension into  the single-user image,  there comes an new problem,   I found the jupyterlab dubugger was missing, I can't even find the bug icon on the my browser. 
I refer to [this doc](https://jupyterlab.readthedocs.io/en/stable/user/debugger.html) to install xeus-python on my EG kernel and refresh browser, but It didn't work, I still can't find the bug icon. 

## Environment
- Jupyter  lab [3.0.11]

**Comment by lresende** [comment]
You should be able to move to the latest IPython kernel (6.x) which now has the support for the debugger. As for the Elyra side, you should be able to use elyra image directly which should have all Elyra components installed.

**Comment by kevin-bates** [comment]
> I refer to this doc to install xeus-python on my EG kernel and refresh browser, but It didn't work, I still can't find the bug icon.

EG's [`ipykernel_launcher.py`](https://github.com/jupyter/enterprise_gateway/blob/d30cf13fc4521b7ff0b594b93a43c70d9c50f334/etc/kernel-launchers/python/scripts/launch_ipykernel.py#L382) embeds `ipykernel` so using xeus-python with EG would require some modifications to the launcher.  As a result, you're probably still launching the ipykernel.  If all you need from the xeus-python kernel is debugger support, then, as Luciano says, upgrading the version of IPython (and ipykernel) in the kernel image may be sufficient.

**Comment by bug-developer021** [comment]
> You should be able to move to the latest IPython kernel (6.x) which now has the support for the debugger. As for the Elyra side, you should be able to use elyra image directly which should have all Elyra components installed.

For the HUB side, I use the [elyra:2.2.4](https://hub.docker.com/layers/elyra/elyra/2.2.4/images/sha256-38941e01b9a82714c4bc7d57b5cf18068a264491afe5846ff901277f8019db89?context=explore) to replace the default single-user image.
And the EG side I use the [elyra/kernel-py:2.5.0](https://hub.docker.com/layers/elyra/kernel-py/2.5.0/images/sha256-9dcafcd99b40a33cd8e58424e35afadeabdd4abd5da453f3ca007ca1843fe983?context=explore) and  upgrade ipykernel to 6.0.3, jupyterlab to 3.0.16.
But I still can't see the debugger icon in my browser
![with_JEG](https://user-images.githubusercontent.com/80243243/126758592-ecfa5730-0411-436b-b117-e560cc73bd34.png)

But when I disconnected JupyterHub with Enterprise Gateway,  the debuuger icon appears in my browser
```yaml
#  extraEnv:
#    JUPYTER_GATEWAY_URL: http://enterprise-gateway.jupyter.svc.cluster.local:8888
#    JUPYTER_GATEWAY_REQUEST_TIMEOUT: "120"
```

**Comment by bug-developer021** [comment]
> > I refer to this doc to install xeus-python on my EG kernel and refresh browser, but It didn't work, I still can't find the bug icon.
> 
> EG's [`ipykernel_launcher.py`](https://github.com/jupyter/enterprise_gateway/blob/d30cf13fc4521b7ff0b594b93a43c70d9c50f334/etc/kernel-launchers/python/scripts/launch_ipykernel.py#L382) embeds `ipykernel` so using xeus-python with EG would require some modifications to the launcher. As a result, you're probably still launching the ipykernel. If all you need from the xeus-python kernel is debugger support, then, as Luciano says, upgrading the version of IPython (and ipykernel) in the kernel image may be sufficient.

Yeah, I try to use the elyra image and upgrade ipykernel to 6.0.3,  jupyterlab to 3.0.16，But I still can't see debugger icon unless I disconnected JupyterHub with Enterprise Gateway

**Comment by kevin-bates** [comment]
Hmm - I don't know what's involved with the addition of debugger support, but there might be something it depends on that is not available through the gateway integration in the notebook server.  Lab 3.0 - in this particular configuration, is talking to Jupyter Server (not Notebook) - which is then configured to redirect its kernel management to EG.  EG, however, is still built on Notebook (server).  I'd be really surprised something related to kernel debugging is required on the server that manages the kernel.  (I figured debugging leverages the message protocol entirely - so that should just work.)

Does debugging require any kind of server extension?

@lresende - do you know what else might be required for ipykernel debugging?

**Comment by kevin-bates** [comment]
OK.  I've found what is involved with configuring ipykernel for debugging.  There is no issue with EG other than its configuration is referencing out-of-date configurations.  The following items are necessary:

1. `ipykernel` version >= 6.0.  This is only available on Python versions >= 3.7.
2. The addition of a `"debugger": true` entry in the `metadata` stanza of the applicable `kernel.json` files (i.e., those associated with `ipykernel`)

So the following needs to happen...

1. The ipykernel kernelspecs baked into the EG image need their `kernel.json` files updated.
2. The Docker file for the `kernel-py` image needs to have its [`BASE_CONTAINER` reference](https://github.com/jupyter/enterprise_gateway/blob/master/etc/docker/kernel-py/Dockerfile#L2) updated to a more recent scipy-notebook image.  This base container is already running an appropriate version of Python.

Because this issue has morphed into a completely separate topic from above - yet holds good discussion information - I'd like to re-open this issue and update the title.  I will also edit the description to indicate its evolution.

**Comment by bug-developer021** [comment]
> OK. I've found what is involved with configuring ipykernel for debugging. There is no issue with EG other than its configuration is referencing out-of-date configurations. The following items are necessary:
> 
> 1. `ipykernel` version >= 6.0.  This is only available on Python versions >= 3.7.
> 2. The addition of a `"debugger": true` entry in the `metadata` stanza of the applicable `kernel.json` files (i.e., those associated with `ipykernel`)
> 
> So the following needs to happen...
> 
> 1. The ipykernel kernelspecs baked into the EG image need their `kernel.json` files updated.
> 2. The Docker file for the `kernel-py` image needs to have its [`BASE_CONTAINER` reference](https://github.com/jupyter/enterprise_gateway/blob/master/etc/docker/kernel-py/Dockerfile#L2) updated to a more recent scipy-notebook image.  This base container is already running an appropriate version of Python.
> 
> Because this issue has morphed into a completely separate topic from above - yet holds good discussion information - I'd like to re-open this issue and update the title. I will also edit the description to indicate its evolution.
> 
> @bug-developer021 - is this something you'd like to contribute? I would be happy to guide you through the process if you need assistance.

Thanks again!  After I set  `"debugger": true` in the `metadata` stanza of the  remote `kernel.json` files,  It sloved my  problem, I can use debugger from JHUB to EG!

**Comment by kevin-bates** [comment]
Hi @bug-developer021 - the changes would involve:

1. Updating the python-related [kernelspecs](https://github.com/kevin-bates/enterprise_gateway/tree/master/etc/kernelspecs) to include the `"debugger": true` entry in their `kernel.json` files.
2. Checking the `FROM` image statement of the python-related [docker images](https://github.com/jupyter/enterprise_gateway/tree/master/etc/docker) that they include `ipykernel >= 6.0`.  This is probably an exercise in updating the `tag` of the source image found in [docker-stacks](https://github.com/jupyter/docker-stacks) to a more recent version.  The tag corresponds to a commit hash in the repo.

I see that we issue `conda install ipykernel` in some of the images.  Since that apparently updates to more recent versions, those images are probably fine as-is.

**Comment by kevin-bates** [comment]
Closed via #986
